feature(files): adds a service for serving files from filestore

hypeJunction opened the pull request:
Adds API and handlers for serving files from the filestore
- [x] How to handle file entities with custom filestore?
- [x] Write docs
- [x] Write tests
- [ ] Address comment https://github.com/Elgg/Elgg/pull/9147#discussion_r48444138

hypeJunction committed:
feature(files): adds a service for serving files from filestore

Adds API and handlers for serving files from the filestore

hypeJunction commented:
This looks quite nice, I think. Let me know what you think, and we can work on cleaning this up, adding tests and docs. We should also avoid booting engine fully for this. Perhaps we should add an option in `Elgg\Application::bootCore($boot_plugins = true)`?

mrclay commented on engine/classes/Elgg/FileService/File.php:
```diff
@@ -0,0 +1,127 @@
+	 */
+	public function setDisposition($disposition = self::DISPOSITION_ATTACHMENT) {
+		if (!in_array($disposition, array(self::DISPOSITION_ATTACHMENT, self::DISPOSITION_INLINE))) {
+			throw new \UnexpectedValueException("Disposition $disposition is not supported in " . __CLASS__);
```
InvalidArgumentException

mrclay commented on engine/classes/Elgg/FileService/File.php:
```diff
@@ -0,0 +1,127 @@
+		);
+
+
+		$key = _elgg_services()->siteSecret->get();
```
Use `get(true)` (binary key preferred over ASCII encoding). Or just use `elgg_build_hmac($data)->getToken()` which does this.

mrclay commented:
Time to replace some core URLs with this API? I don't think we could merge without some real testing.

hypeJunction commented:
Yes, will make some time tomorrow. I want this in badly.
On Nov 26, 2015 12:11 AM, "Steve Clay" <notifications@github.com> wrote:

> Time to replace some core URLs with this API? I don't think we could merge
> without some real testing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/9147#issuecomment-159752364>.
>

juho-jaakkola commented on engine/lib/filestore.php:
```diff
@@ -581,6 +581,43 @@ function _elgg_filestore_test($hook, $type, $value) {
+ * @param string    $expires    URL expiration, as a string suitable for strtotime()
+ * @return string
+ */
+function elgg_get_inline_url(\ElggFile $file, $use_cookie = false, $expires = '+1 year') {
```
I wonder which is better API:

```php
elgg_get_inline_url($file, false, '+1 year');
```
or
```php
elgg_get_url(array(
    'file' => $file,
    'use_cookie' => false,
    'expires' => '+1 year',
    'disposition' => 'inline',
));
```

hypeJunction replied:
I think the idea is to not make devs think about it too much and let core handle the internals.

juho-jaakkola commented on engine/classes/Elgg/Application/ServeFileHandler.php:
```diff
@@ -0,0 +1,179 @@
+	 * @return void
+	 */
+	protected function handle304($etag) {
+		// If is the same ETag, content didn't change.
```
```If ETag is the same, content hasn't changed.```

This old comment has bugged me more than it should. :)

hypeJunction commented:
File icons and downloads, as well as user avatars are handled with the new API.
I have also made the following change:
 * we need to validate that mtime of the file matches the one in the URL
 * we need to set default expire to 0 for inline URLs so that the URL structure is the same on consecutive requests and we can benefit from cache
 * fixed bugs with etag, so we now get proper 304 headers

hypeJunction commented on engine/classes/Elgg/Application/ServeFileHandler.php:
```diff
@@ -0,0 +1,183 @@
+		$this->application->bootCore();
+
+
+		$key = _elgg_services()->siteSecret->get();
```
Remove, no longer needed.

hypeJunction commented on engine/classes/Elgg/Application/ServeFileHandler.php:
```diff
@@ -0,0 +1,183 @@
+		if ($expires) {
+			$expires_str = gmdate('D, d M Y H:i:s \G\M\T', $expires);
+		} else {
+			$expires_str = gmdate('D, d M Y H:i:s \G\M\T', strtotime("+3 years"));
```
Should we use a different default?

mrclay replied:
I always see 1 year. Doesn't matter at all :)

hypeJunction commented on engine/classes/Elgg/Application/ServeFileHandler.php:
```diff
@@ -0,0 +1,183 @@
+
+		header('Expires: ' . $expires_str, true);
+
+		$cache_control = $use_cookie ? 'no-cache' : 'public';
```
Do we need to add max-age?

mrclay replied:
Expires should be sufficient, but for HTTP/1.1, yeah let's add it (only on Cache-Control, not Pragma)

hypeJunction replied:
Hm, I don't think I understand cache control enough. I will let it sit for now, and we can add it later.

hypeJunction commented on engine/classes/Elgg/Application/ServeFileHandler.php:
```diff
@@ -0,0 +1,183 @@
+
+		header("ETag: $etag");
+
+		readfile($filenameonfilestore);
```
Should we add `X-Sendfile` if Apache and `mod_xsendfile` is enabled?

mrclay replied:
Let's tackle that later. Already a ticket for it I bet.

hypeJunction commented on engine/classes/Elgg/FileService/File.php:
```diff
@@ -0,0 +1,126 @@
+
+	const DISPOSITION_INLINE = 'inline';
+	const DISPOSITION_ATTACHMENT = 'attachment';
+	const DEFAULT_TTL = 7200;
```
No longer using this

hypeJunction commented on engine/classes/Elgg/FileService/File.php:
```diff
@@ -0,0 +1,126 @@
+ */
+class File {
+
+	const DISPOSITION_INLINE = 'inline';
```
Should we shorten to `INLINE` and `ATTACHMENT`?

mrclay replied:
Sure.

hypeJunction commented on mod/file/start.php:
```diff
@@ -359,66 +359,86 @@ function file_set_url($hook, $type, $url, $params) {
+			$readfile->setFilename($thumbfile);
+			$thumb_url = elgg_get_inline_url($readfile, true);
+			if ($thumb_url) {
+				return $thumb_url;
```
We don't have an icon for simpletype = image. Need to fallback to default entity icon?

mrclay replied:
What's the current code do in this case?

mrclay replied:
We can just file a bug and deal later.

hypeJunction replied:
It always return a URL, which then silently fails.

hypeJunction commented on mod/file/start.php:
```diff
@@ -359,66 +359,86 @@ function file_set_url($hook, $type, $url, $params) {
+				break;
 		}
 
-		$mapping = array(
```
I terminated early, and this is just indentation here.

hypeJunction commented on mod/file/thumbnail.php:
```diff
@@ -1,10 +1,10 @@
 <?php
+
```
I say we deprecate this file and get rid of it in 3.x

mrclay replied:
Sure.

hypeJunction commented on mod/file/views/default/resources/file/download.php:
```diff
@@ -15,23 +15,4 @@
-flush();
-readfile($file->getFilenameOnFilestore());
-exit;
+forward(elgg_get_download_url($file));
```
Deprecate this view?

mrclay replied:
:+1:

hypeJunction commented on views/default/resources/avatar/view.php:
```diff
@@ -12,39 +12,13 @@
+	$filehandler = new ElggFile();
+	$filehandler->owner_guid = $user->guid;
+	$filehandler->setFilename("profile/{$user->guid}{$size}.jpg");
+	$avatar_url = elgg_get_inline_url($filehandler);
```
This can return `false`

hypeJunction commented on engine/classes/Elgg/Application/ServeFileHandler.php:
```diff
@@ -0,0 +1,183 @@
+			$this->send403('URL has expired');
+		}
+
+		$mime = $this->getContentType($path_from_dataroot);
```
this should be `$filenameonfilestore`

mrclay commented on mod/profile/start.php:
```diff
@@ -107,34 +103,7 @@ function profile_set_url($hook, $type, $url, $params) {
-		elgg_log("Unable to get profile icon for user with GUID $user_guid", 'ERROR');
-		return elgg_get_simplecache_url("icons/default/$size.png");
-	}
+	elgg_deprecated_notice("Profile plugin no longer customizes avatar url using 'entity:icon:url' hook", '2.1');
```
Good!

hypeJunction commented:
Re: files with custom filestore. I don't think file plugin is capable of handling such cases anyway, and anyone implementing custom filestore would have to rewrite all handlers and use hooks to implement custom serving logic.
Re: tests. What is the strategy? I don't think we can actually test the ServeFileHandler. FileService\File would probably just testing a bunch of combinations and asserting that the actual URL matches and expected URL?

hypeJunction commented on engine/classes/Elgg/FileService/File.php:
```diff
@@ -0,0 +1,134 @@
+			$relative_path,
+		);
+
+		return elgg_normalize_url(implode('/', $url_segments));
```
The line was too long erroring Travis tests.

juho-jaakkola commented on engine/classes/Elgg/Application/ServeFileHandler.php:
```diff
@@ -0,0 +1,180 @@
+	protected function handle304($etag) {
+		if (isset($this->server_vars['HTTP_IF_NONE_MATCH'])
+			&& trim($this->server_vars['HTTP_IF_NONE_MATCH']) === $etag) {
+			header("HTTP/1.1 304 Not Modified");
```
I guess tests could use ```headers_list()``` for testing these ```handleNNN()``` methods.

```$this->assertContains("HTTP/1.1 304 Not Modified", headers_list());```

juho-jaakkola replied:
Or we could inject ```Symfony\Component\HttpFoundation\Response``` and mock that in the tests.

hypeJunction replied:
Yes, I think we should inject and mock `Symfony\Component\HttpFoundation\Response`. We can then eliminate `exit()` calls.

hypeJunction committed:
chore(docs): file system documentation

hypeJunction commented:
@mrclay @juho-jaakkola I have rewritten the server to use Symfony's `Response` and `BinaryFileResponse`, have added some tests and docs. I am not sure it makes sense to test the `File` separately, as most of it is tested in `ServeFileHandler`.

hypeJunction commented:
Test failures seem unrelated. Would be good if someone could check.

mrclay commented on mod/file/actions/file/upload.php:
```diff
@@ -68,6 +68,10 @@
 		$title = $file->title;
 	}
+
+	if ($entity->access_id !== $access_id) {
```
The changes in this file look like a bugfix that could be moved to a separate PR for 2.0, or am I missing something? Also you'd want to set `$reset_icon_urls` either way so you don't get notices below.

juho-jaakkola replied:
I agree.

hypeJunction replied:
These changes would make no sense in 2.0. They only make sense for icons handled by the new API. What happens here is we update icons' modified times, so all requests to previously generated URLs will fail due to HMAC mismatch.

juho-jaakkola commented on engine/tests/phpunit/Elgg/Application/ServeFileHandlerTest.php:
```diff
@@ -0,0 +1,166 @@
+	/**
+	 * @group FileService
+	 */
+	public function testResponseWhenFileModifiedTime() {
```
Test names should describe the expected result. E.g. `testSends403onFileModificationTimeMismatch` or similar.

juho-jaakkola replied:
Same for the other tests below.

mrclay commented:
Strawman: Make a separate download handler like file/download for files with non-default filestore (we'd need a method to determine this). They'd use separate URL that contains GUID instead of path.

hypeJunction commented:
I think we should get the API in first, and update existing server logic in separate PRs. I will move the 2 commits into separate PRs and add something in developer tools for testing.

hypeJunction commented:
I have stripped out the changes to avatars and file thumbs. Those are in separate PRs now.
@juho-jaakkola Updated test names.

hypeJunction committed:
fix(files): files with custom filestore can now be served via file service

Adds the logic to serve files by their GUID with full application boot, if they have
a custom filestore

hypeJunction commented:
Added `DownloadFileHandler` for files that have custom filestore. It starts the application, retrieves the file and serves it as `attachment`

mrclay commented on engine/classes/Elgg/FileService/File.php:
```diff
@@ -0,0 +1,142 @@
+		if (!$relative_path) {
+			// File object has a custom filestore
+			if ($this->file->guid) {
+				$url_segments = array(
```
This API will let you request a public URL with expires and inline, but depending on the file object, it just won't obey you; you'll get a download link that requires login. We should give URLs that elicit the same behavior. I'll try to send a PR to do that in the next few days.

hypeJunction replied:
Ok. I can use a break from this feature :baby_bottle:

juho-jaakkola replied:
The problem is related only to custom filestores? One option could be to keep using the old URLs for bundled plugins, deprecate support for custom filestores during 2.x, and remove completely in Elgg 3.

juho-jaakkola replied:
@mrclay Can you elaborate the problem you've found?

mrclay replied:
Use the API to generate a public URL that's served inline. What you instead get is a download link that requires login.

Your suggestion is simplest: Kill custom filestores in 3.0 and add this then.

juho-jaakkola replied:
I'm in fact suggesting that we add the API already in 2.x, but we wait until Elgg 3 to use it in bundled plugins, so we don't break any sites using custom filestore.

mrclay commented:
This will also resolve #5891.

juho-jaakkola commented:
Anyone against including this in Elgg 2.1 and adding a warning that it won't work with custom filestores? 

Or did @mrclay already have some idea how to make this work with them?

mrclay commented:
Yeah, let's pull this in for 2.1. See also #9352.

mrclay commented:
Replaced by #9354
